### ISSUE TITLE
attempt to index field 'weight' (a nil value)

### ISSUE THREAD

**Issue by burnzzz** [facebookresearch/fairseq-lua#5]
If I try to run this example I get an exeption:

`fairseq generate-lines -path wmt14.en-de.fconv-cuda/model.th7 -sourcedict wmt14.en-de.fconv-cuda/dict.de.th7 -targetdict wmt14.en-de.fconv-cuda/dict.en.th7 -beam 5
`
| [target] Dictionary: 42242 types
| [source] Dictionary: 43675 types
> Dieser Zug fährt vom München nach Paris
/home/xxx/tools/torch/install/bin/luajit: ...rch/install/share/lua/5.1/fairseq/models/fconv_model.lua:84: attempt to index field 'weight' (a nil value)
stack traceback:
        ...rch/install/share/lua/5.1/fairseq/models/fconv_model.lua:84: in function 'onEvaluate'
        ...install/share/lua/5.1/fairseq/modules/TrainTestLayer.lua:29: in function 'evaluate'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        ...el/tools/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        ...rch/install/share/lua/5.1/fairseq/models/fconv_model.lua:652: in function 'setup'
        .../install/share/lua/5.1/fairseq/models/ensemble_model.lua:102: in function 'generate'
        ...install/share/lua/5.1/fairseq/scripts/generate-lines.lua:218: in main chunk
        [C]: in function 'require'
        ...rch/install/lib/luarocks/rocks/fairseq/scm-1/bin/fairseq:17: in main chunk
        [C]: at 0x00405d50

Any ideas?

**Comment by jgehring** [comment]
Please try updating your nn installation, e.g. by doing `luarocks install nn`.

**Comment by michaelauli** [comment]
Did the suggestion above help?

**Comment by burnzzz** [comment]
I first updated torch:
`luarocks install torch`

then I updated nn
`luarocks install nn`

then I tried again and now get:

` fairseq generate-lines -path wmt14.en-de.fconv-cuda/model.th7 -sourcedict wmt14.en-de.fconv-cuda/dict.de.th7 -targetdict wmt14.en-de.fconv-cuda/dict.en.th7 -beam 5
| [target] Dictionary: 42242 types
| [source] Dictionary: 43675 types
> Dieser Zug fährt vom München nach Paris
/home/daniel/tools/torch/install/bin/luajit: .../install/share/lua/5.1/fairseq/models/ensemble_model.lua:134: bad argument #2 to 'add' (sizes do not match at /tmp/luarocks_cutorch-scm-1-8598/cutorch/lib/THC/generated/../generic/THCTensorMathPointwise.cu:216)
stack traceback:
        [C]: in function 'add'
        .../install/share/lua/5.1/fairseq/models/ensemble_model.lua:134: in function 'generate'
        ...install/share/lua/5.1/fairseq/scripts/generate-lines.lua:218: in main chunk
        [C]: in function 'require'
        ...rch/install/lib/luarocks/rocks/fairseq/scm-1/bin/fairseq:17: in main chunk
        [C]: at 0x00405d50
`

That´s progress, right?

**Comment by jgehring** [comment]
Almost there :) This is an English to German model, so `-sourcedict` should be English, i.e. `dict.en.th7` and `-targetdict` should be German.

**Comment by schani** [comment]
I have the same problem, and I did as @burnzzz did, first update `torch`, then `nn`.  If I just update `nn` on a Torch installation from the latest commit I get compile errors.

Still, having updated `torch` and `nn`, the problem still persists:

```
ubuntu@BorisGPUMP2 [ /vol/fairseq ] 
[18:29]: env CUDA_VISIBLE_DEVICES=3 fairseq generate-lines -path wmt14.en-fr.fconv-cuda/model.th7 -sourcedict wmt14.en-fr.fconv-cuda/dict.en.th7 \
    -targetdict wmt14.en-fr.fconv-cuda/dict.fr.th7 -beam 5
| [target] Dictionary: 44666 types
| [source] Dictionary: 44409 types
> The cat sat on the mat.
/home/ubuntu/torch/install/bin/luajit: ...rch/install/share/lua/5.1/fairseq/models/fconv_model.lua:84: attempt to index field 'weight' (a nil value)
stack traceback:
        ...rch/install/share/lua/5.1/fairseq/models/fconv_model.lua:84: in function 'onEvaluate'
        ...install/share/lua/5.1/fairseq/modules/TrainTestLayer.lua:29: in function 'evaluate'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'func'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:73: in function 'applyToModules'
        /home/ubuntu/torch/install/share/lua/5.1/nn/Container.lua:91: in function 'evaluate'
        ...rch/install/share/lua/5.1/fairseq/models/fconv_model.lua:652: in function 'setup'
        .../install/share/lua/5.1/fairseq/models/ensemble_model.lua:102: in function 'generate'
        ...install/share/lua/5.1/fairseq/scripts/generate-lines.lua:218: in main chunk
        [C]: in function 'require'
        ...rch/install/lib/luarocks/rocks/fairseq/scm-1/bin/fairseq:17: in main chunk
        [C]: at 0x00405d50
```

These are the relevant entries from `luarocks list`

```
torch
   scm-1 (installed) - /home/ubuntu/torch/install/lib/luarocks/rocks
nn
   scm-1 (installed) - /home/ubuntu/torch/install/lib/luarocks/rocks
```

**Comment by jgehring** [comment]
@schani, this still looks like the error that you'd get with an older version of nn. Can you quickly check `/home/ubuntu/torch/install/share/lua/5.1/nn/WeightNorm.lua` and compare it with https://github.com/torch/nn/blob/master/WeightNorm.lua, in particular whether your local version features the function at the very bottom called `read`?

**Comment by schani** [comment]
You're right, it's not the last version. Trying to figure out what to do about it...

**Comment by burnzzz** [comment]
It works now. Like @jgehring said, the sourcedict and targetdict have to fit the model... 

`fairseq generate-lines -path wmt14.en-de.fconv-cuda/model.th7 -sourcedict wmt14.en-de.fconv-cuda/dict.en.th7 -targetdict wmt14.en-de.fconv-cuda/dict.de.th7 -beam 5
| [target] Dictionary: 43675 types
| [source] Dictionary: 42242 types
> quit
S       <unk> 
O       quit
H       -0.91358071565628       - Herr Präsident ! 
A       1 1 1 1 1
> 
`

Interesting translation for "quit" :)

**Comment by jgehring** [comment]
Hehe, well to be fair "quit" is an unknown word and the training data contains speeches from the European Parliament :) With the BPE codes in the model directory, we can get the proper input (the translation is still wrong, though):

```
S	qu@@ it
O	qu@@ it
H	-0.12036862969398	qu@@ it
A	1 2 3
```

**Comment by schani** [comment]
I got it to work, finally with Torch installed according to [their website](http://torch.ch/docs/getting-started.html).  I used current `master` on `torch/distro` (`9e2f875d0ee3d82e16d87b5abc0d64c92b50fd10`), and had to use commit `3752f2426b55bc32cbd0ef112649d47dc674baa8` in `extra/nn`.  It's not sufficient to just check out that commit in the submodule directory because the Torch install script will check out whatever submodule commit is committed in the `distro` repo, so I had to commit the submodule checkout.

This is my commit: https://github.com/schani/distro/commit/ffd04326093738a0d552f411318c065857351c67
